proposals/0569-multiline-strings.rst

```diff
@@ -0,0 +1,556 @@
+
+Costs and Drawbacks
+-------------------
+
+Since this only affects lexing and parsing, I expect development and maintenance costs to be low. This feature is common in other languages, so there shouldn't be any learning curve for new developers coming from another language. If anything, the auto-stripping of leading whitespace might be a source of confusion, but a one-line explanation should be sufficient.
+
+Alternatives
+------------
+
+* Status quo, e.g. using ``unlines``
+
+  * As mentioned in the Motivation, it's not great ergonomics, but it works.
+
+* Third party libraries, using quasiquoters
+
+  * Template Haskell is overkill for this
+
+* No stripping of leading whitespace
+
+  * This probably comes from one of two concerns: more complex implementation, conceptually adds automagic. It does make the implementation a bit harder, but this is a small enough change that I don't think it makes the overall proposal much harder to implement. While it does add a bit more magic behind the scenes, I think the rule is simple enough (no more complex than do-block indentation rules) and the use-case common enough (I can't think of a single use-case that would want the indentation to be part of the string) that it warrants the bump in ergonomics.
+
+  * The downside of doing this is that generally speaking, developers will want to keep the multiline string at the same indentation level as surrounding code. Not doing any post processing means that reindenting code would change the string content. I would also posit that the common case is wanting leading whitespace stripped, which would lead to devs putting multiline strings at the 0th column or implementing their own deindenter.
+
+* Use ``''`` to delimit multiline strings, which has the benefit of being a parse error without ``MultilineStrings``
+
+* Enable any number of ``"""+`` quotes to delimit multiline strings
+
+* Reuse single-quoted ``"`` for multiline syntax
+
+  * Would require escaping double quotes in the multiline string, which, while not a major part of the proposal, is a nice bonus
+
+* New ``[s|foo|]`` construct that embeds a multiline literal string with no TH (so *not* a quasiquoter, but reusing the same syntax)
+
+* Support arbitrary terminators, like Bash's heredocs
+
+  ::
+
+    x = <<EOF
+    line 1
+    line 2
+    line 3
+    EOF
+
+  * Everyone will use a different terminator, which I think would contribute to a reduction in overall readability
+  * I think ``"""`` is an uncommon enough delimiter, and it can be escaped, that I don't think this is necessary
+
+* Use some delimiter to start the string, but use layout indentation rules to dictate when the string ends
+
+* Strip trailing whitespace in post-processing
+
+  * Nice to have, but not necessary. I think it would be better to keep post-processing as minimal as possible, and it doesn't seem as common as removing leading whitespace.
+
+Out of scope
```

Comment:
I'm fine with these being out of scope for this proposal, but I would add a few sentences here to sketch out how this proposal could be forwards-compatible with interpolation and raw strings.